pkg/library/status/check.go

```diff
@@ -0,0 +1,167 @@
+//
+// Copyright (c) 2019-2022 Red Hat, Inc.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/devfile/devworkspace-operator/pkg/common"
+	"github.com/devfile/devworkspace-operator/pkg/config"
+	"github.com/devfile/devworkspace-operator/pkg/infrastructure"
+	"github.com/devfile/devworkspace-operator/pkg/provision/sync"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var containerFailureStateReasons = []string{
+	"CrashLoopBackOff",
+	"ImagePullBackOff",
+	"CreateContainerError",
+	"RunContainerError",
+}
+
+// unrecoverablePodEventReasons contains Kubernetes events that should fail workspace startup
+// if they occur related to a workspace pod. Events are stored as a map with event names as keys
+// and values representing the threshold of how many times we can see an event before it is considered
+// unrecoverable.
+var unrecoverablePodEventReasons = map[string]int32{
+	"FailedPostStartHook":   1,
+	"FailedMount":           3,
+	"FailedScheduling":      1,
+	"FailedCreate":          1,
+	"ReplicaSetCreateError": 1,
+}
+
+var unrecoverableDeploymentConditionReasons = []string{
+	"FailedCreate",
+}
+
+func CheckDeploymentStatus(deployment *appsv1.Deployment) (ready bool) {
+	return deployment.Status.ReadyReplicas > 0
+}
+
+func CheckDeploymentConditions(deployment *appsv1.Deployment) (healthy bool, errorMsg string) {
+	conditions := deployment.Status.Conditions
+	for _, condition := range conditions {
+		for _, unrecoverableReason := range unrecoverableDeploymentConditionReasons {
+			if condition.Reason == unrecoverableReason {
+				return false, fmt.Sprintf("Detected unrecoverable deployment condition: %s %s", condition.Reason, condition.Message)
+			}
+		}
+	}
+	return true, ""
+}
+
+// checkPodsState checks if workspace-related pods are in an unrecoverable state. A pod is considered to be unrecoverable
+// if it has a container with one of the containerFailureStateReasons states, or if an unrecoverable event (with reason
+// matching unrecoverablePodEventReasons) has the pod as the involved object.
+// Returns optional message with detected unrecoverable state details
+//         error if any happens during check
+func CheckPodsState(workspaceID string, namespace string, labelSelector k8sclient.MatchingLabels,
```

Comment:
Since you updated this to take a label selector, it's almost identical to `checkCleanupPodsState` -- why not just call this instead of `checkCleanupPodsState` and wrap a non-empty message with `Common PVC cleanup job failed: %s`?